STAR.py

```diff
@@ -1,11 +1,145 @@
 import pandas as pd
 import numpy as np
 
-def STAR(K, W, S):
+class InvalidElection(Exception):
+    pass
+
+class TrueTie:
+    def __init__(self, tied):
+        self.tied = tied
+
+def summary_data(ballots: pd.DataFrame):
+    # The below quantities can be emitted however is most useful
+    score_hist = {a : ballots[a].value_counts() for a in ballots.columns}
+    score_sums = ballots.sum()
+    pairwise_matrix = { a :
+            { b : pairwise_winner(ballots, a, b) for b in ballots.columns }
+        for a in ballots.columns
+    }
+    return (score_hist, score_sums, pairwise_matrix)
+
+def score_winners(ballots: pd.DataFrame):
     #Score winners
-    df_tops = S[S.sum().nlargest(2, keep='all').index]
+    scores = ballots.sum()
+    top_score = scores.max()
+    return [a for a in ballots.columns if np.isclose(scores[a], top_score)]
+    
+def pairwise_winner(a: pd.Series, b: pd.Series):
+    # Returns 1 if a is preferred to b, -1 if b is preferred to a, and 0 for a tie
+    margin = (a > b).sum() - (b > a).sum()
+    if margin > 0:
+        return 1
+    elif margin < 0:
+        return -1
+    else:
+        return 0
+
+def has_defeat(ballots: pd.DataFrame, a):
+    for b in ballots.columns:
+        if pairwise_winner(ballots[b], ballots[a]) == 1:
+            return 1
+    return 0
+
+def fivestar(ballots: pd.DataFrame):
+    fstars = (ballots == ballots.max().max()).sum()
+    top_fstars = fstars.max()
+    return [a for a in ballots.columns if np.isclose(fstars[a], top_fstars)]
+
+def weak_condorcet_winners(ballots: pd.DataFrame):
+    return [a for a in ballots.columns if not has_defeat(ballots, a)]
+
+##############################################################################
+# The above functions are mainly helpers for easier reading of the main logic.
+##############################################################################
+
+def default_scoring_tiebreaker(ballots: pd.DataFrame):
+    tmp = weak_condorcet_winners(ballots)
+    if len(tmp) == 1:
+        return tmp[0]
+    elif len(tmp) == 0:
+        return TrueTie(list(ballots.columns))
+    else:
+        return TrueTie(tmp)
+
+def default_runoff_tiebreaker(ballots: pd.DataFrame):
+    # Any pre-runoff ties must be resolved
+    assert len(ballots.columns) == 2
+
+    a,b = ballots.columns
+    scores = ballots.sum()
+    if np.isclose(scores[a], scores[b]):
+        return TrueTie(list(ballots.columns))
+    else:
+        return scores.argmax()
+
+def STAR(ballots: pd.DataFrame, scoring_tiebreaker=default_scoring_tiebreaker, runoff_tiebreaker=default_runoff_tiebreaker):
```

Comment:
Thanks for updating the variable names, and adding the docstring. It's a lot easier to read now